### PR TITLE
fix memory leak

### PR DESCRIPTION
**Pull request by riyazhakki**
#### Bug/feature Description
* *Memory allocated in mom_process_hook is not freed*

#### Affected Platform(s)
* *All*

#### Cause / Analysis / Design
* *Malloc memory in mom_process_hook was not freed*

#### Solution Description
* *Releasing malloced memory in mom_process_hook before returning from the function*

#### Testing logs/output
* *Please attach your test log output from running the test you added (or from existing tests that cover your changes)*

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/resmom/mom_server.c [hunk 1]
```diff
@@ -1102,6 +1102,8 @@ is_request(int stream, int version)
 							log_err(errno, __func__, MALLOC_ERR_MSG);
 							goto err;
 					}
+					memset(phook_output->reject_errcode, 0, sizeof(int));
+
```
Comment (mike0042) [inline src/resmom/mom_server.c]: This might be more than needed. How about... phook_output->reject_errcode = 0;

Comment (riyazhakki) [reply src/resmom/mom_server.c]: Thank you Mike, modified the statement.

File: src/resmom/mom_hook_func.c [hunk 2]
```diff
@@ -3517,7 +3518,7 @@ void reply_hook_bg(job *pjob)
 	return;
 
 	fini:
-		/*free(php->hook_output->reject_errcode);*/
+		free(php->hook_output->reject_errcode);
```
Comment (bayucan) [inline src/resmom/mom_hook_func.c]: Do you really need to free this? This is declared   in include/mom_hook_func.h as:
typedef struct mom_hook_output {
        int             *reject_errcode;
I

Comment (riyazhakki) [reply src/resmom/mom_hook_func.c]: We are doing malloc for this at below lines
[mom_server.c#L1101](https://github.com/PBSPro/pbspro/blob/master/src/resmom/mom_server.c#L1101) , [requests.c#L626](https://github.com/PBSPro/pbspro/blob/master/src/resmom/requests.c#L626)

Comment (bayucan) [reply src/resmom/mom_hook_func.c]: I see the malloc() part now. Thanks. So yes, it's okay to free() here.

Comment (bayucan) [reply src/resmom/mom_hook_func.c]: Now in mom_server.c, beginning in line 1094, we have this: 

					if ((phook_output = (mom_hook_output_t *)malloc(
						sizeof(mom_hook_output_t))) == NULL) {
							log_err(errno, __func__, MALLOC_ERR_MSG);
							goto err;
					}
					mom_hook_output_init(phook_output);

					if ((phook_output->reject_errcode =
						(int *)malloc(sizeof(int))) == NULL) {
							log_err(errno, __func__, MALLOC_ERR_MSG);
							goto err;
					}
If malloc() of reject_errcode fails, we also need to free up the malloc-ed phook_output in the previous statement. I looked at the 'err' part, and I don't see a line where phook_output is freed. And only when mom_process_hooks() is called, does the phook->output is put on a task.

Similar situation in requests.c just before the call to EXECJOB_END, we have:
619         if ((hook_output = (mom_hook_output_t *)malloc(
 620                 sizeof(mom_hook_output_t))) == NULL) {
 621                         log_err(errno, __func__, MALLOC_ERR_MSG);
 622                         return;
 623         }
 624         mom_hook_output_init(hook_output);
 625         if ((hook_output->reject_errcode =
 626                 (int *)malloc(sizeof(int))) == NULL) {
 627                         log_err(errno, __func__, MALLOC_ERR_MSG);
 628                         return;
 629         }
line 628 early return needs to free(hook_output)...

Comment (riyazhakki) [reply src/resmom/mom_hook_func.c]: Thank you Al, for catching this. I have addressed it.

**Comment by riyazhakki** [comment]
Attaching test logs
[TestPbsExecjobEnd.ptl.txt](https://github.com/PBSPro/pbspro/files/2893595/TestPbsExecjobEnd.ptl.txt)
[TestPbsExecjobEnd_MOM1.val.txt](https://github.com/PBSPro/pbspro/files/2893596/TestPbsExecjobEnd_MOM1.val.txt)
[TestPbsExecjobEnd_MOM2.val.txt](https://github.com/PBSPro/pbspro/files/2893597/TestPbsExecjobEnd_MOM2.val.txt)

Running few more tests

**Comment by riyazhakki** [comment]
Sorry for the late on this PR. I am done with testing and fixed an invalid free, please review the same.

**Comment by prakashcv13** [comment]
merged #996

**Comment by riyazhakki** [comment]
Thank you Prakash.

**Pull request merged**